lib/ui/painting.dart

```diff
@@ -1316,6 +1316,26 @@ Future<Null> _decodeImageFromListAsync(Uint8List list, ImageDecoderCallback call
   callback(frameInfo.image);
 }
 
+/// The encoding formats supported by the [encodeImage].
+enum EncodedImageFormat {
+  // Be conservative with the formats we expose. We can increase the list in future but reducing
+  // it is more difficult.
+  JPEG,
+  PNG,
+  WEBP,
+}
+
+/// Callback signature for [encodeImage].
+typedef void ImageEncoderCallback(Uint8List result);
+
+/// Convert an [Image] object into a byte array/
+///
+/// [format] is encoding format to be used.
+///
+/// [quality] is a value in [0, 100] where 0 corresponds to the lowest quality.
+void encodeImage(Image image, EncodedImageFormat format, int quality, ImageEncoderCallback callback)
```

Comment:
This should be a method on Image.

Reply:
We should probably make the format and quality arguments named optional arguments with defaults.
It's probably also a good idea to have this return a `Future<ByteData>` rather than using the callback approach. That would be more idiomatic. We can use the callback approach internally in this function in order to implement the Future, though, since dealing with Futures and the engine is non-trivial otherwise.

Reply:
Also, if we make this a method on Image, we should probably call it `toByteData`, to be more idiomatic.

Reply:
+1.  FYI for context, see https://github.com/flutter/flutter/issues/11648#issuecomment-350902741

Reply:
Ack. I think it makes sense for this to be a method on Image. 

Can you please provide my an example of how a Future returning method is implemented in
engine?
I presume Hixie's suggestion was to have a private method implemented natively that uses
callback and then call this private method from public dart method. Effectively creating
the future in dart and resolving that future via callback roughly:

``` 
class Image {
  Future<Uint8List> toByteData(format, quality) {
      final Completer<Uint8List> completer = new Completer<Uint8List>();
      _encodeImage(this, format, quality, completer.complete); // this is natively implemented.
     return completer.future;
   }
}
```

Reply:
Yup, that's it exactly.

Reply:
You can just use the `_futurize` method in this file to do it for you though.